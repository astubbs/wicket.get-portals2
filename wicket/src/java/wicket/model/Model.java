/*
 * $Id$ $Revision:
 * 1.11 $ $Date$
 * 
 * ==============================================================================
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package wicket.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import wicket.Component;

/**
 * Model is the basic implementation of an AbstractModel. It just wraps a simple
 * model object. The model object must be serializable, as it is stored in the
 * session. If you have large objects to store, consider using
 * {@link AbstractDetachableModel}instead of this class.
 * 
 * @author Chris Turner
 * @author Eelco Hillenius
 */
public class Model<V /* extends Serializable*/> extends AbstractModel<V>
{
	private static final long serialVersionUID = 1L;

	/** Backing object. */
	private V object;

	/**
	 * Construct the model without providing an object.
	 */
	public Model()
	{
	}

	/**
	 * Construct the model, setting the given object as the wrapped object.
	 * 
	 * @param object
	 *            The model object proper
	 */
	public Model(final V object)
	{
		setObject(object);
	}

	/**
	 * @param map
	 *            The Map, which may or may not be Serializable
	 * @return A Model object wrapping the Map
	 */
	public static Model<Map> valueOf(final Map map)
	{
		return new Model<Map>(map instanceof Serializable ? map : new HashMap(map));
	}

	/**
	 * @param list
	 *            The List, which may or may not be Serializable
	 * @return A Model object wrapping the List
	 */
	public static Model<Serializable> valueOf(final List list)
	{
		return new Model<Serializable>(list instanceof Serializable ? (Serializable)list : new ArrayList(list));
	}

	/**
	 * @see wicket.model.IModel#getNestedModel()
	 */
	@Override
	public IModel getNestedModel()
	{
		return null;
	}

	/**
	 * @see wicket.model.IModel#getObject(wicket.Component)
	 */
	public V getObject(final Component component)
	{
		return object;
	}

	/**
	 * Set the model object; calls setObject(java.io.Serializable). The model
	 * object must be serializable, as it is stored in the session
	 * 
	 * @param object
	 *            the model object
	 * @see wicket.model.IModel#setObject(Component, Object)
	 */
	public void setObject(final Component component, final V object)
	{
		setObject(object);
	}

	/**
	 * Sets the model object. The model object must be serializable, as it is
	 * stored in the session
	 * 
	 * @param object
	 *            The serializable model object
	 * @see wicket.model.IModel#setObject(Component, Object)
	 */
	public void setObject(final V object)
	{
		this.object = object;
	}

	/**
	 * @see java.lang.Object#toString()
	 */
	@Override
	public String toString()
	{
		StringBuffer sb = new StringBuffer(super.toString());
		sb.append(":nestedModel=[").append(getNestedModel()).append("]");
		sb.append(":object=[").append(this.object).append("]");
		return sb.toString();
	}
}
