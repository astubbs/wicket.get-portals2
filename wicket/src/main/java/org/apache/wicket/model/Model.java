/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.wicket.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.wicket.Component;
import org.apache.wicket.WicketRuntimeException;


/**
 * <code>Model</code> is the basic implementation of an <code>IModel</code>. It just wraps a
 * simple model object. The model object must be serializable, as it is stored in the session. If
 * you have large objects to store, consider using
 * {@link org.apache.wicket.model.LoadableDetachableModel} instead of this class.
 * 
 * @author Chris Turner
 * @author Eelco Hillenius
 * 
 * @param <T>
 *            The Model Object
 */
public class Model<T extends Serializable> implements IModel<T>
{
	private static final long serialVersionUID = 1L;

	/** Backing object. */
	private T object;

	/**
	 * Construct the model without providing an object.
	 */
	public Model()
	{
	}

	/**
	 * Construct the model, setting the given object as the wrapped object.
	 * 
	 * @param object
	 *            The model object proper
	 */
	public Model(final T object)
	{
		setObject(object);
	}

	/**
	 * @param map
	 *            The Map, which may or may not be Serializable
	 * @deprecated see {@link Model#ofMap(Map)}
	 * @return A Model object wrapping the Map
	 */
	@Deprecated
	public static Model valueOf(final Map map)
	{
		return new Model(map instanceof Serializable ? (Serializable)map : new HashMap(map));
	}

	/**
	 * @param list
	 *            The List, which may or may not be Serializable
	 * @deprecated see {@link Model#of(List)}
	 * @return A Model object wrapping the List
	 */
	@Deprecated
	public static Model valueOf(final List list)
	{
		return new Model(list instanceof Serializable ? (Serializable)list : new ArrayList(list));
	}

	/**
	 * Factory method for models that contain maps. This factory method will automatically rebuild a
	 * nonserializable <code>map</code> into a serializable one.
	 * 
	 * @param <K>
	 *            key type in map
	 * @param <V>
	 *            value type in map
	 * @param <T>
	 *            model type
	 * @param map
	 *            The Map, which may or may not be Serializable
	 * @return A Model object wrapping the Map
	 */
	@SuppressWarnings("unchecked")
	public static <K, V, T extends Map<K, V> & Serializable> Model<T> of(final Map<K, V> map)
	{
		return new Model<T>((T)(map instanceof Serializable ? map : new HashMap<K, V>(map)));
	}

	/**
	 * Factory method for models that contain lists. This factory method will automatically rebuild
	 * a nonserializable <code>list</code> into a serializable one.
	 * 
	 * @param <V>
	 *            value type in list
	 * @param <T>
	 *            model type
	 * @param list
	 *            The List, which may or may not be Serializable
	 * @return A Model object wrapping the List
	 */
	@SuppressWarnings("unchecked")
	public static <V, T extends List<V> & Serializable> Model<T> of(final List<V> list)
	{
		return new Model<T>((T)(list instanceof Serializable ? list : new ArrayList<V>(list)));
	}

	/**
	 * Factory method for models that contain sets. This factory method will automatically rebuild a
	 * nonserializable <code>set</code> into a serializable one.
	 * 
	 * @param <V>
	 *            value type in set
	 * @param <T>
	 *            model type
	 * @param set
	 *            The Set, which may or may not be Serializable
	 * @return A Model object wrapping the Set
	 */
	@SuppressWarnings("unchecked")
	public static <V, T extends Set<V> & Serializable> Model<T> of(final Set<V> set)
	{
		return new Model<T>((T)(set instanceof Serializable ? set : new HashSet<V>(set)));
	}


	/**
	 * Factory methods for Model which uses type inference to make code shorter. Eqivalent to
	 * <code>new Model<TypeOfObject>(object)</code>.
	 * 
	 * @param <T>
	 * @param object
	 * @return Model that contains <code>object</code>
	 */
	public static <T extends Serializable> Model<T> of(T object)
	{
		return new Model<T>(object);
	}

	/**
	 * @see org.apache.wicket.model.IModel#getObject()
	 */
	public T getObject()
	{
		return object;
	}

	/**
	 * Set the model object; calls setObject(java.io.Serializable). The model object must be
	 * serializable, as it is stored in the session
	 * 
	 * @param object
	 *            the model object
	 * @see org.apache.wicket.model.IModel#setObject(Object)
	 */
	public void setObject(final T object)
	{
		if (object != null)
		{
			if (!(object instanceof Serializable))
			{
				throw new WicketRuntimeException("Model object must be Serializable");
			}
		}
		this.object = object;
	}

	/**
	 * @see org.apache.wicket.model.IDetachable#detach()
	 */
	public void detach()
	{
		if (object instanceof IDetachable)
		{
			((IDetachable)object).detach();
		}
	}

	/**
	 * @see java.lang.Object#toString()
	 */
	@Override
	public String toString()
	{
		StringBuffer sb = new StringBuffer("Model:classname=[");
		sb.append(getClass().getName()).append("]");
		sb.append(":object=[").append(object).append("]");
		return sb.toString();
	}

	// TODO These methods are for helping people upgrade. Remove after
	// deprecation release.
	/**
	 * @param component
	 * @return
	 * @deprecated replace by {@link IModel#getObject()}.
	 */
	@Deprecated
	public final Object getObject(Component component)
	{
		throw new UnsupportedOperationException();
	}

	/**
	 * @param component
	 * @param object
	 * @deprecated replace by {@link IModel#setObject(Object)}.
	 */
	@Deprecated
	public final void setObject(Component component, Object object)
	{
		throw new UnsupportedOperationException();
	}
}
