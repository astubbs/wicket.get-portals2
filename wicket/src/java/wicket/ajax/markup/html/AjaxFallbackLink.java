/*
 * $Id: AjaxFallbackLink.java 4633 2006-02-25 16:22:21 -0800 (Sat, 25 Feb 2006)
 * dashorst $ $Revision$ $Date: 2006-02-25 16:22:21 -0800 (Sat, 25 Feb
 * 2006) $
 * 
 * ==============================================================================
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package wicket.ajax.markup.html;

import wicket.MarkupContainer;
import wicket.ajax.AjaxEventBehavior;
import wicket.ajax.AjaxRequestTarget;
import wicket.ajax.IAjaxCallDecorator;
import wicket.ajax.calldecorator.CancelEventIfNoAjaxDecorator;
import wicket.markup.html.link.Link;
import wicket.model.IModel;

/**
 * An ajax link that will degrade to a normal request if ajax is not available
 * or javascript is disabled
 * 
 * @since 1.2
 * 
 * @author Igor Vaynberg (ivaynberg)
 * 
 */
public abstract class AjaxFallbackLink extends Link implements IAjaxLink
{
	/** */
	private static final long serialVersionUID = 1L;

	/**
	 * Construct.
	 * 
	 * @param parent
	 *            The parent of this component
	 * 
	 * @param id
	 */
	public AjaxFallbackLink(MarkupContainer parent, final String id)
	{
		this(parent, id, null);
	}

	/**
	 * Construct.
	 * 
	 * @param parent
	 *            The parent of this component
	 * 
	 * @param id
	 * @param model
	 */
	public AjaxFallbackLink(MarkupContainer parent, final String id, final IModel model)
	{
		super(parent, id, model);

		add(new AjaxEventBehavior("onclick")
		{
			private static final long serialVersionUID = 1L;

			@Override
			protected void onEvent(AjaxRequestTarget target)
			{
				onClick(target);
			}

			@Override
			protected IAjaxCallDecorator getAjaxCallDecorator()
			{
				return new CancelEventIfNoAjaxDecorator(AjaxFallbackLink.this
						.getAjaxCallDecorator());
			}

		});
	}

	protected IAjaxCallDecorator getAjaxCallDecorator()
	{
		return null;
	}

	/**
	 * 
	 * @see wicket.markup.html.link.Link#onClick()
	 */
	@Override
	public final void onClick()
	{
		onClick(null);
	}

	/**
	 * Callback for the onClick event. If ajax failed and this event was
	 * generated via a normal link the target argument will be null
	 * 
	 * @param target
	 *            ajax target if this linked was invoked using ajax, null
	 *            otherwise
	 */
	public abstract void onClick(final AjaxRequestTarget target);
}
