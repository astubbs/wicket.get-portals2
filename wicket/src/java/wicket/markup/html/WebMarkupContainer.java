/*
 * $Id$
 * $Revision$ $Date$
 * 
 * ==============================================================================
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package wicket.markup.html;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import wicket.Component;
import wicket.HtmlHeaderContainer;
import wicket.MarkupContainer;
import wicket.markup.MarkupElement;
import wicket.markup.MarkupException;
import wicket.markup.MarkupStream;
import wicket.markup.WicketTag;
import wicket.model.IModel;
import wicket.util.lang.Classes;

/**
 * A container of HTML markup and components. It is very similar to the base
 * class MarkupContainer, except that the markup type is defined to be HTML.
 * 
 * @author Jonathan Locke
 */
public class WebMarkupContainer extends MarkupContainer
{
    private transient List headerComponents;
    
	/**
	 * @see Component#Component(String)
	 */
	public WebMarkupContainer(final String id)
	{
		super(id);
	}
	
	/**
	 * @see wicket.Component#Component(String, IModel)
	 */
	public WebMarkupContainer(final String id, IModel model)
	{
		super(id, model);
	}

	/**
	 * Gets the markup type for this component.
	 * 
	 * @return Markup type of HTML
	 */
	public final String getMarkupType()
	{
		return "html";
	}

	/**
	 * Renders this component. This implementation just calls renderComponent.
	 */
	protected void onRender()
	{
		renderComponent(findMarkupStream());
	}
	
	
	/**
	 * Gets the header part for this markup container. Returns null if it doesn't contribute
	 * a header part.
	 * @param index TODO
	 * @return the header part for this markup container or null
	 * 	if it doesn't contribute anything.
	 */
	public WebMarkupContainer getHeaderPart(int index)
	{
	    // gracefull getAssociateMarkupStream. Throws no exception in case
	    // markup is not found
		final MarkupStream associatedMarkupStream = 
		    	getApplication().getMarkupCache().getMarkupStream(this, null, false);
		
		if (associatedMarkupStream == null)
		{
		    return null;
		}

		do
		{
		    final MarkupElement element = associatedMarkupStream.get();
		    if (element instanceof WicketTag)
		    {
		        final WicketTag wTag = (WicketTag) element;
		        if (wTag.isHeadTag() == true)
		        {
		            final String headerId = "_" + Classes.name(this.getClass()) + "Header";
		            WebMarkupContainer headerContainer = new HtmlHeaderContainer(headerId, associatedMarkupStream);
		            if (this.headerComponents != null)
		            {
		                for (Iterator iter = headerComponents.iterator(); iter.hasNext(); )
		                {
		                    headerContainer.add((Component) iter.next());
		                }
		                
		                // Cleanup; no longer needed.
		                this.headerComponents = null;
		            }
		            return headerContainer;
		        }
		    }
		} 
		while (associatedMarkupStream.next() != null);

		if (this.headerComponents == null)
		{
		    throw new MarkupException("You have added header components but did not specific a <wicket:head> region in your markup");
		}
    	return null;
	}
	
	/**
	 * Add component to &lt;wicket:head&gt; instead of panel region. 
	 * 
	 * @param child
	 */
	public void addToHeader(final Component child)
	{
	    if (this.headerComponents == null)
	    {
	        this.headerComponents = new ArrayList();
	    }
	    
	    this.headerComponents.add(child);
	}
}