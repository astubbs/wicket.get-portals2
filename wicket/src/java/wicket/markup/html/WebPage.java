/*
 * $Id$ $Revision$
 * $Date$
 * 
 * ==============================================================================
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package wicket.markup.html;

import java.util.Iterator;

import wicket.ApplicationPages;
import wicket.Component;
import wicket.Page;
import wicket.PageMap;
import wicket.PageParameters;
import wicket.WicketRuntimeException;
import wicket.markup.html.ajax.IAjaxListener;
import wicket.markup.html.link.BookmarkablePageLink;
import wicket.markup.parser.filter.HtmlHeaderSectionHandler;
import wicket.model.IModel;
import wicket.protocol.http.WebRequest;
import wicket.protocol.http.WebRequestCycle;
import wicket.util.lang.Classes;
import wicket.util.string.Strings;

/**
 * Base class for HTML pages. This subclass of Page simply returns HTML when
 * asked for its markup type. It also has a method which subclasses can use to
 * retrieve a bookmarkable link to the application's home page.
 * <p>
 * WebPages can be constructed with any constructor when they are being used in
 * a Wicket session, but if you wish to link to a Page using a URL that is
 * "bookmarkable" (which implies that the URL will not have any session
 * information encoded in it, and that you can call this page directly without
 * having a session first directly from your browser), you need to implement
 * your Page with a no-arg constructor or with a constructor that accepts a
 * PageParameters argument (which wraps any query string parameters for a
 * request). In case the page has both constructors, the constructor with
 * PageParameters will be used.
 * 
 * @author Jonathan Locke
 * @author Eelco Hillenius
 * @author Juergen Donnerstag
 * @author Gwyn Evans
 */
public class WebPage extends Page implements IHeaderRenderer
{
	private static final long serialVersionUID = 1L;

	/** Components contribution to <body onload="..." */
	private String bodyOnLoad;

	/**
	 * Constructor. Having this constructor public means that you page is
	 * 'bookmarkable' and hence can be called/ created from anywhere.
	 */
	protected WebPage()
	{
		super();
	}

	/**
	 * Constructor which receives wrapped query string parameters for a request.
	 * Having this constructor public means that you page is 'bookmarkable' and
	 * hence can be called/ created from anywhere. For bookmarkable pages (as
	 * opposed to when you construct page instances yourself, this constructor
	 * will be used in preference to a no-arg constructor, if both exist. Note
	 * that nothing is done with the page parameters argument. This constructor
	 * is provided so that tools such as IDEs will include it their list of
	 * suggested constructors for derived classes.
	 * 
	 * @param parameters
	 *            Wrapped query string parameters.
	 */
	protected WebPage(final PageParameters parameters)
	{
		this();
	}

	/**
	 * @see Page#Page(IModel)
	 */
	protected WebPage(final IModel model)
	{
		super(model);
	}

	/**
	 * Returns a bookmarkable URL that references a given page class using a
	 * given set of page parameters. Since the URL which is returned contains
	 * all information necessary to instantiate and render the page, it can be
	 * stored in a user's browser as a stable bookmark.
	 * 
	 * @param pageMapName
	 *            Name of pagemap to use
	 * @param pageClass
	 *            Class of page
	 * @param parameters
	 *            Parameters to page
	 * @return Bookmarkable URL to page
	 */
	public String urlFor(final String pageMapName, final Class pageClass,
			final PageParameters parameters)
	{
		if (pageClass == null)
		{
			throw new NullPointerException("argument pageClass may not be null");
		}

		final WebRequestCycle cycle = getWebRequestCycle();
		final StringBuffer buffer = urlPrefix(cycle);
		if (pageMapName == null)
		{
			appendPageMapName(buffer);
		}
		else
		{
			buffer.append("?pagemap=");
			buffer.append(pageMapName);
			buffer.append('&');
		}
		ApplicationPages pages = getApplicationPages();

		// "bookmarkablePage=xxx" is required if PageParameters exist,
		// some sort of redirection takes place, or if we're not dealing
		// with the homepage.
		if ((parameters != null && !parameters.isEmpty())
				|| pages.getHomePageRenderStrategy() != ApplicationPages.NO_REDIRECT
				|| !pages.getHomePage().equals(pageClass))
		{
			buffer.append("bookmarkablePage=");
			String pageReference = cycle.getApplication().getPages().aliasForClass(pageClass);
			if (pageReference == null)
			{
				pageReference = pageClass.getName();
			}
			buffer.append(pageReference);
		}
		if (parameters != null)
		{
			for (final Iterator iterator = parameters.keySet().iterator(); iterator.hasNext();)
			{
				final String key = (String)iterator.next();
				final String value = parameters.getString(key);
				if (value != null)
				{
					final String escapedValue = Strings.escapeMarkup(value);
					buffer.append('&');
					buffer.append(key);
					buffer.append('=');
					buffer.append(escapedValue);
				}
			}
		}
		if (buffer.charAt(buffer.length() - 1) == '?')
		{
			buffer.deleteCharAt(buffer.length() - 1);
		}
		return cycle.getResponse().encodeURL(buffer.toString());
	}

	/**
	 * Returns a URL that references a given interface on a component. When the
	 * URL is requested from the server at a later time, the interface will be
	 * called. A URL returned by this method will not be stable across sessions
	 * and cannot be bookmarked by a user.
	 * 
	 * @param component
	 *            The component to reference
	 * @param listenerInterface
	 *            The listener interface on the component
	 * @return A URL that encodes a page, component and interface to call
	 */
	public String urlFor(final Component component, final Class listenerInterface)
	{
		// Ensure that component instanceof listenerInterface
		if (!listenerInterface.isAssignableFrom(component.getClass()))
		{
			throw new WicketRuntimeException("The component " + component + " of class "
					+ component.getClass() + " does not implement " + listenerInterface);
		}

		// Buffer for composing URL
		final WebRequestCycle cycle = getWebRequestCycle();
		final StringBuffer buffer = urlPrefix(cycle);
		appendPageMapName(buffer);
		buffer.append("path=");
		buffer.append(component.getPath());
		int versionNumber = component.getPage().getCurrentVersionNumber();
		if (versionNumber > 0)
		{
			buffer.append("&version=");
			buffer.append(versionNumber);
		}
		String listenerName = Classes.name(listenerInterface);
		if (!"IRedirectListener".equals(listenerName))
		{
			buffer.append("&interface=");
			buffer.append(listenerName);
		}

		// add an extra parameter for regconition in case we are targetting a
		// dispatched handler
		if (IAjaxListener.class.isAssignableFrom(listenerInterface))
		{
			buffer.append("&dispatched=true");
		}

		return cycle.getResponse().encodeURL(buffer.toString());
	}

	/**
	 * @param path
	 *            The path
	 * @return The url for the path
	 */
	public String urlFor(final String path)
	{
		if ((path == null) || (path.trim().length() == 0))
		{
			return urlPrefix(getWebRequestCycle()).toString();
		}
		else
		{
			return urlPrefix(getWebRequestCycle()) + "/" + path;
		}
	}

	/**
	 * Gets the markup type for a WebPage, which is "html" by default. Support
	 * for pages in another markup language, such as VXML, would require the
	 * creation of a different Page subclass in an appropriate package under
	 * wicket.markup. To support VXML (voice markup), one might create the
	 * package wicket.markup.vxml and a subclass of Page called VoicePage.
	 * <p>
	 * Note: The markup type must be equal to the extension of the markup file.
	 * In the case of WebPages, it must always be "html".
	 * 
	 * @return Markup type for HTML
	 */
	public String getMarkupType()
	{
		return "html";
	}

	/**
	 * @return The WebRequestCycle for this WebPage.
	 */
	protected final WebRequestCycle getWebRequestCycle()
	{
		return (WebRequestCycle)getRequestCycle();
	}

	/**
	 * Creates and returns a bookmarkable link to this application's home page.
	 * 
	 * @param id
	 *            Name of link
	 * @return Link to home page for this application
	 */
	protected final BookmarkablePageLink homePageLink(final String id)
	{
		return new BookmarkablePageLink(id, getApplicationPages().getHomePage());
	}

	/**
	 * Appends any pagemap name to the buffer
	 * 
	 * @param buffer
	 *            The string buffer to append to
	 */
	private void appendPageMapName(final StringBuffer buffer)
	{
		final PageMap pageMap = getPageMap();
		if (!pageMap.isDefault())
		{
			buffer.append("?pagemap=");
			buffer.append(pageMap.getName());
			buffer.append('&');
		}
		else
		{
			buffer.append('?');
		}
	}

	/**
	 * Creates a prefix for a url.
	 * 
	 * @param cycle
	 *            The web request cycle
	 * @return Prefix for URLs including the context path and servlet path.
	 */
	private StringBuffer urlPrefix(final WebRequestCycle cycle)
	{
		final StringBuffer buffer = new StringBuffer();
		final WebRequest request = cycle.getWebRequest();
		if (request != null)
		{
			final String contextPath = request.getContextPath();
			buffer.append(contextPath);
			String path = request.getServletPath();
			if (path == null || "".equals(path))
			{
				path = "/";
			}
			buffer.append(path);
		}

		return buffer;
	}

	/**
	 * THIS IS NOT PART OF WICKETS PUBLIC API. DO NOT USE IT YOURSELF.
	 * <p>
	 * Invoked by HtmlHeaderContainer it'll ask all child components of the Page
	 * if they have something to contribute to the &lt;head&gt; section of the
	 * HTML output. Every component interested must implement
	 * IHeaderContributor.
	 * <p>
	 * Note: HtmlHeaderContainer will be removed from the component hierachie at
	 * the end of the request (@see #onEndRequest()) and thus can not transport
	 * status from one request to the next. This is true for all components
	 * added to the header.
	 * 
	 * @see IHeaderContributor
	 * @param container
	 *            The header component container
	 */
	public final void renderHeaderSections(final HtmlHeaderContainer container)
	{
		// A components interested in contributing to the header must
		// implement IHeaderContributor.
		visitChildren(IHeaderContributor.class, new IVisitor()
		{
			/**
			 * @see wicket.Component.IVisitor#component(wicket.Component)
			 */
			public Object component(Component component)
			{
				if (component.isVisible())
				{
					if (component instanceof IHeaderContributor)
					{
						((IHeaderContributor)component).renderHead(container);
					}
				}
				return IVisitor.CONTINUE_TRAVERSAL;
			}
		});
	}

	/**
	 * THIS IS NOT PART OF THE PUBLIC API.
	 * 
	 * Get what will be appended to the page markup's body onload attribute
	 * 
	 * @return The onload attribute
	 */
	public String getBodyOnLoad()
	{
		return this.bodyOnLoad;
	}

	/**
	 * THIS IS NOT PART OF THE PUBLIC API.
	 * 
	 * Append string to body onload attribute
	 * 
	 * @param onLoad
	 *            Attribute value to be appended
	 */
	public final void appendToBodyOnLoad(final String onLoad)
	{
		if (onLoad != null)
		{
			// Tell the page to change the Page's
			// body tags.
			if (this.bodyOnLoad == null)
			{
				this.bodyOnLoad = onLoad;
			}
			else
			{
				this.bodyOnLoad = this.bodyOnLoad + onLoad;
			}
		}
	}

	/**
	 * Remove the header component and all its children from the component
	 * hierachie. Be aware, thus you can not transfer state from one request to
	 * another.
	 * 
	 * @see wicket.Component#onEndRequest()
	 */
	protected void onEndRequest()
	{
		final Component header = get(HtmlHeaderSectionHandler.HEADER_ID);
		if (header != null)
		{
			this.remove(header);
		}

		super.onEndRequest();
	}
}