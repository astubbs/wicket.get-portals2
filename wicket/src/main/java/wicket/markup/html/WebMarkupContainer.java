/*
 * $Id: WebMarkupContainer.java 5861 2006-05-25 20:55:07 +0000 (Thu, 25 May
 * 2006) eelco12 $ $Revision$ $Date: 2006-05-25 20:55:07 +0000 (Thu, 25
 * May 2006) $
 * 
 * ==============================================================================
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package wicket.markup.html;

import wicket.Component;
import wicket.MarkupContainer;
import wicket.model.IModel;

/**
 * A container of HTML markup and components. It is very similar to the base
 * class MarkupContainer, except that the markup type is defined to be HTML.
 * 
 * @param <T>
 *            Type of model object this component holds
 * 
 * @author Jonathan Locke
 * @author Juergen Donnerstag
 */
public class WebMarkupContainer<T> extends MarkupContainer<T>
{
	private static final long serialVersionUID = 1L;

	/**
	 * @see Component#Component(MarkupContainer,String)
	 */
	public WebMarkupContainer(MarkupContainer parent, final String id)
	{
		super(parent, id);
	}

	/**
	 * @see wicket.Component#Component(MarkupContainer,String, IModel)
	 */
	public WebMarkupContainer(MarkupContainer parent, final String id, IModel<T> model)
	{
		super(parent, id, model);
	}

	/**
	 * Gets the markup type for this component.
	 * 
	 * @return Markup type of HTML
	 */
	@Override
	public final String getMarkupType()
	{
		return "html";
	}

	/**
	 * A convinience method to return the WebPage. Same as getPage().
	 * 
	 * @return WebPage
	 */
	public final WebPage getWebPage()
	{
		return (WebPage)getPage();
	}
}