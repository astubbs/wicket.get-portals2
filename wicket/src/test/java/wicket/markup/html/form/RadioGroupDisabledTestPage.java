/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package wicket.markup.html.form;

import wicket.markup.html.WebMarkupContainer;
import wicket.markup.html.WebPage;
import wicket.model.Model;

/**
 * Tests rendering of the RadioGroup and Radio components
 * @author igor
 */
public class RadioGroupDisabledTestPage extends WebPage
{
	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;

	
	/**
	 * Constructor
	 */
	@SuppressWarnings("unchecked") // It's not the warnings which are important here.
	public RadioGroupDisabledTestPage() {
		Form form=new Form(this, "form");
		RadioGroup group=new RadioGroup(form, "group", new Model("radio2"));
		group.setRenderBodyOnly(false);
		WebMarkupContainer container=new WebMarkupContainer(group, "container");
		Radio radio1=new Radio(group, "radio1", new Model("radio1"));
		Radio radio2=new Radio(container, "radio2", new Model("radio2"));
		
		group.setEnabled(false);
	}
}
