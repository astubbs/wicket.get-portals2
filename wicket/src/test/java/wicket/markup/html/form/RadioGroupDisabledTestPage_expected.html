<html>
<head><title>RadioGroupTestPage1</title></head>
<body>
	<form action="/WicketTester/WicketTester?wicket:interface=:0:form::IFormSubmitListener" wicket:id="form" method="post" id="form0"><div style="display:none"><input type="hidden" name="form0_hf_0" id="form0_hf_0" /></div>
		<span disabled="disabled" wicket:id="group" name="group">
			<input value="radio1" type="radio" disabled="disabled" wicket:id="radio1" name="group">radio1</input>
			<span wicket:id="container">
				<input value="radio2" type="radio" disabled="disabled" wicket:id="radio2" checked="checked" name="group">radio2</input>
			</span>
		</span>
	</form>
</body>
</html>
